File: src/Python/sync_proposals.py
"""Sync proposals from Koios → PostgreSQL.

Koios: GET /api/v1/proposal_list
Target: proposals table
"""

import sys
import time

from config import TABLE_COLUMNS, BATCH_SIZE, PROPOSAL_TRIGGERS
from helpers import (
    get_logger, check_env, koios_get, pg_connect, pg_upsert_batch,
    pg_row_count, now_iso, gen_uuid, pg_ensure_proposal_activities_table,
    pg_drop_triggers, pg_recreate_proposal_triggers, pg_query,
)


def sync_proposals(logger=None):
    if logger is None:
        logger = get_logger()

    logger.info("=== Sync: proposals (Koios → PostgreSQL) ===")
    columns = TABLE_COLUMNS["proposals"]

    # 1. Fetch from Koios
    logger.info("[proposals] Fetching from Koios...")
    raw = koios_get("proposal_list")
    logger.info(f"[proposals] Got {len(raw)} proposals from Koios")

    # 2. Transform
    rows = []
    seen_ids = set()
    for p in raw:
        pid = p.get("proposal_id")
        if not pid or pid in seen_ids:
            continue
        seen_ids.add(pid)

        body = (p.get("meta_json") or {}).get("body") or {}

        # Combine abstract, rationale, motivation
        abstract_parts = []
        if body.get("abstract"):
            abstract_parts.append(body["abstract"])
        if body.get("rationale"):
            abstract_parts.append(f"RATIONALE:\n{body['rationale']}")
        if body.get("motivation"):
            abstract_parts.append(f"MOTIVATION:\n{body['motivation']}")

        refs = body.get("references") or []
        first_uri = refs[0].get("uri") if refs and isinstance(refs, list) else None
        authors = (p.get("meta_json") or {}).get("authors") or []
        author_name = authors[0].get("name") if authors else None

        rows.append({
            "id": gen_uuid(),
            "proposal_id": pid,
            "title": body.get("title") or pid,
            "abstract": "\n\n".join(abstract_parts) or None,
            "first_reference_uri": first_uri,
            "author_name": author_name,
            "proposal_index": p.get("proposal_index"),
            "proposal_tx_hash": p.get("proposal_tx_hash"),
            "proposed_epoch": p.get("proposed_epoch"),
            "expiration": p.get("expiration"),
            "proposal_type": p.get("proposal_type"),
            "epoch_no": None,
            "status": None,
            "data_fetched_at": now_iso(),
            "created_at": now_iso(),
            "updated_at": now_iso(),
            "activities_table_created": False,
            "activities_table_name": None,
            "abstract_summary": (body.get("abstract") or "")[:500] or None,
            "slug": None,
        })

    logger.info(f"[proposals] Transformed {len(rows)} rows")

    # 3. Connect
    conn = pg_connect()

    # 4. Find existing proposal_ids
    existing_res = pg_query(conn, "SELECT proposal_id FROM proposals")
    existing_ids = {r[0] for r in existing_res}

    # 5. Split new vs existing
    new_rows = [row for row in rows if row["proposal_id"] not in existing_ids]
    existing_rows = [row for row in rows if row["proposal_id"] in existing_ids]
    logger.info(f"[proposals] {len(new_rows)} new to insert, {len(existing_rows)} existing to update")

    # 6. Drop triggers before insert to prevent orphan ga_* tables
    triggers_dropped = False
    if new_rows:
        logger.info(f"[proposals] Dropping triggers before insert ({len(new_rows)} new)...")
        pg_drop_triggers(conn, "proposals", PROPOSAL_TRIGGERS)
        triggers_dropped = True

    # 7. Insert new proposals (no trigger = no orphan ga_* tables)
    inserted = 0
    for i in range(0, len(new_rows), BATCH_SIZE):
        batch = new_rows[i:i + BATCH_SIZE]
        try:
            count = pg_upsert_batch(
                conn, "proposals", columns, batch,
                conflict_cols=["proposal_id"], do_update=False
            )
            inserted += count
        except Exception as e:
            logger.error(f"[proposals] Error at batch {i}: {e}")
        pct = min(100, (i + len(batch)) * 100 // len(new_rows)) if new_rows else 100
        sys.stdout.write(f"\r  [proposals] Insert {pct}% (PostgreSQL={inserted})")
        sys.stdout.flush()
    logger.info(f"\n[proposals] Inserted {inserted}/{len(new_rows)} new")

    # 8. Update existing proposals (UPDATE, not INSERT → no trigger → no orphan ga_* tables)
    #    Note: do NOT update status — it is managed by sync_epoch (active vs done)
    update_fields = [
        "title", "abstract", "first_reference_uri", "author_name",
        "proposal_index", "proposal_tx_hash", "proposed_epoch", "expiration",
        "proposal_type", "data_fetched_at", "updated_at", "abstract_summary",
    ]
    updated = 0
    for i, row in enumerate(existing_rows):
        try:
            with conn.cursor() as cur:
                set_clause = ", ".join(f"{c} = %s" for c in update_fields)
                values = [row.get(c) for c in update_fields]
                cur.execute(
                    f"UPDATE proposals SET {set_clause} WHERE proposal_id = %s",
                    values + [row["proposal_id"]]
                )
                updated += 1
        except Exception as e:
            logger.error(f"[proposals] Update error for {row['proposal_id']}: {e}")
        if (i + 1) % 25 == 0 or i == len(existing_rows) - 1:
            sys.stdout.write(f"\r  [proposals] Update {i+1}/{len(existing_rows)} (PostgreSQL={updated})")
            sys.stdout.flush()
    logger.info(f"\n[proposals] Updated {updated}/{len(existing_rows)}")

    # 9. Recreate triggers for real-time inserts from the app (only if dropped)
    if triggers_dropped:
        logger.info("[proposals] Recreating triggers...")
        pg_recreate_proposal_triggers(conn, "proposals")
    else:
        logger.info("[proposals] Triggers untouched (no inserts)")

    # 10. Ensure ga_* tables exist for proposals missing them
    logger.info("[proposals] Ensuring ga_* tables for proposals missing them...")
    missing_res = pg_query(
        conn,
        "SELECT proposal_id FROM proposals "
        "WHERE activities_table_name IS NULL OR activities_table_created = FALSE"
    )
    missing = [r[0] for r in missing_res]

    created = 0
    for idx, pid in enumerate(missing):
        try:
            pg_ensure_proposal_activities_table(conn, pid)
            created += 1
        except Exception as e:
            logger.error(f"[proposals] Failed to create ga_* table for {pid}: {e}")
        sys.stdout.write(f"\r  [proposals] tables {idx+1}/{len(missing)} created={created}")
        sys.stdout.flush()
    logger.info(f"\n[proposals] Created {created}/{len(missing)} ga_* tables")

    # 11. Verify
    count = pg_row_count(conn, "proposals")
    logger.info(f"[proposals] Row count: {count}")
    conn.close()
    return count


if __name__ == "__main__":
    check_env()
    sync_proposals()

File: src/JavaScript/sync_proposals.js
// sync_proposals.js
// Sync proposals from Koios → PostgreSQL
// Equivalent to sync_proposals.py
//
// Koios: GET /api/v1/proposal_list
// Target: proposals table

require('dotenv').config();
const fetch = require('node-fetch');
const crypto = require('crypto');
const { pgConnect, pgUpsertBatch, info, warn, error } = require('./helpers');
const { TABLE_COLUMNS } = require('./config');

const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) {
  console.error('❌ Thiếu DATABASE_URL trong .env');
  process.exit(1);
}

const KOIOS_BASE = 'https://api.koios.rest/api/v1';
const BATCH_SIZE = 25;
const PROPOSAL_TRIGGERS = [
  'trg_create_proposal_activities_table',
  'trg_create_proposal_summary_entry',
];

const COLUMNS = TABLE_COLUMNS.proposals;

async function jsonGet(url, headers = {}) {
  const res = await fetch(url, { headers });
  if (!res.ok) throw new Error(`${url} → HTTP ${res.status}`);
  return res.json();
}

async function koiosGet(endpoint) {
  return jsonGet(`${KOIOS_BASE}/${endpoint}`, { Accept: 'application/json' });
}

function genUuid() {
  return crypto.randomUUID();
}

function nowIso() {
  return new Date().toISOString();
}

async function syncProposals() {
  info('=== Sync: proposals (Koios → PostgreSQL) ===');

  const client = await pgConnect();

  try {
    // 1. Fetch from Koios
    info('[proposals] Fetching from Koios...');
    const raw = await koiosGet('proposal_list');
    info(`[proposals] Got ${raw.length} proposals from Koios`);

    // 2. Transform
    const seenIds = new Set();
    const rows = [];
    for (const p of raw) {
      const pid = p.proposal_id;
      if (!pid || seenIds.has(pid)) continue;
      seenIds.add(pid);

      const body = (p.meta_json || {}).body || {};
      const abstractParts = [];
      if (body.abstract) abstractParts.push(body.abstract);
      if (body.rationale) abstractParts.push(`RATIONALE:\n${body.rationale}`);
      if (body.motivation) abstractParts.push(`MOTIVATION:\n${body.motivation}`);

      const refs = body.references || [];
      const firstUri = refs.length > 0 && refs[0].uri ? refs[0].uri : null;
      const authors = (p.meta_json || {}).authors || [];
      const authorName = authors.length > 0 ? authors[0].name : null;

      rows.push({
        id: genUuid(),
        proposal_id: pid,
        title: body.title || pid,
        abstract: abstractParts.length > 0 ? abstractParts.join('\n\n') : null,
        first_reference_uri: firstUri,
        author_name: authorName,
        proposal_index: p.proposal_index,
        proposal_tx_hash: p.proposal_tx_hash,
        proposed_epoch: p.proposed_epoch,
        expiration: p.expiration,
        proposal_type: p.proposal_type,
        epoch_no: null,
        status: null,
        data_fetched_at: nowIso(),
        created_at: nowIso(),
        updated_at: nowIso(),
        activities_table_created: false,
        activities_table_name: null,
        abstract_summary: (body.abstract || '').slice(0, 500) || null,
        slug: null,
      });
    }
    info(`[proposals] Transformed ${rows.length} rows`);

    // 3. Find existing proposal_ids
    const existingRes = await client.query('SELECT proposal_id FROM proposals');
    const existingIds = new Set(existingRes.rows.map(r => r.proposal_id));

    // 4. Split new vs existing
    const newRows = rows.filter(row => !existingIds.has(row.proposal_id));
    const existingRows = rows.filter(row => existingIds.has(row.proposal_id));
    info(`[proposals] ${newRows.length} new to insert, ${existingRows.length} existing to update`);

    // 5. Drop triggers before insert (to prevent orphan ga_* tables)
    let triggersDropped = false;
    if (newRows.length > 0) {
      info(`[proposals] Dropping triggers before insert (${newRows.length} new)...`);
      for (const tg of PROPOSAL_TRIGGERS) {
        await client.query(
          `DROP TRIGGER IF EXISTS ${tg} ON proposals`
        );
      }
      triggersDropped = true;
    }

    // 6. Insert new proposals (no trigger = no orphan ga_* tables)
    let inserted = 0;
    for (let i = 0; i < newRows.length; i += BATCH_SIZE) {
      const batch = newRows.slice(i, i + BATCH_SIZE);
      try {
        inserted += await pgUpsertBatch(client, 'proposals', COLUMNS, batch, {
          conflict_cols: ['proposal_id'],
          do_update: false,
        });
      } catch (e) {
        error(`[proposals] Error at batch ${i}: ${e.message}`);
      }
      const pct = Math.min(100, Math.round(((i + batch.length) * 100) / newRows.length));
      process.stdout.write(`\r  [proposals] Insert ${pct}% (PostgreSQL=${inserted})`);
    }
    info(`\n[proposals] Inserted ${inserted}/${newRows.length} new`);

    // 7. Update existing proposals (UPDATE, not INSERT → no trigger)
    const updateFields = [
      'title', 'abstract', 'first_reference_uri', 'author_name',
      'proposal_index', 'proposal_tx_hash', 'proposed_epoch', 'expiration',
      'proposal_type', 'data_fetched_at', 'updated_at', 'abstract_summary',
    ];
    let updated = 0;
    for (let i = 0; i < existingRows.length; i++) {
      const row = existingRows[i];
      try {
        const setClause = updateFields.map(c => `${c} = $${updateFields.indexOf(c) + 1}`).join(', ');
        const values = updateFields.map(c => row[c]);
        await client.query(
          `UPDATE proposals SET ${setClause} WHERE proposal_id = $${updateFields.length + 1}`,
          [...values, row.proposal_id]
        );
        updated++;
      } catch (e) {
        error(`[proposals] Update error for ${row.proposal_id}: ${e.message}`);
      }
      if ((i + 1) % 25 === 0 || i === existingRows.length - 1) {
        process.stdout.write(`\r  [proposals] Update ${i + 1}/${existingRows.length} (PostgreSQL=${updated})`);
      }
    }
    info(`\n[proposals] Updated ${updated}/${existingRows.length}`);

    // 8. Recreate triggers
    if (triggersDropped) {
      info('[proposals] Recreating triggers...');
      await client.query(`
        CREATE TRIGGER trg_create_proposal_activities_table
        BEFORE INSERT ON proposals FOR EACH ROW
        EXECUTE FUNCTION create_proposal_activities_table()
      `);
      await client.query(`
        CREATE TRIGGER trg_create_proposal_summary_entry
        AFTER INSERT ON proposals FOR EACH ROW
        EXECUTE FUNCTION create_proposal_summary_entry()
      `);
    } else {
      info('[proposals] Triggers untouched (no inserts)');
    }

    // 9. Ensure ga_* tables for proposals missing them
    info('[proposals] Ensuring ga_* tables for proposals missing them...');
    const missingRes = await client.query(
      "SELECT proposal_id FROM proposals WHERE activities_table_name IS NULL OR activities_table_created = FALSE"
    );
    const missing = missingRes.rows.map(r => r.proposal_id);

    let created = 0;
    for (let idx = 0; idx < missing.length; idx++) {
      const pid = missing[idx];
      try {
        const sanitized = pid.replace(/[^a-zA-Z0-9]/g, '').slice(0, 40);
        const tableName = `ga_${crypto.createHash('md5').update(pid).digest('hex').slice(0, 10)}_${sanitized}`;
        const md5Short = crypto.createHash('md5').update(pid).digest('hex').slice(0, 8);

        await client.query(`
          CREATE TABLE IF NOT EXISTS "${tableName}" (
            id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
            block_time TIMESTAMP WITH TIME ZONE NULL,
            voter_role VARCHAR(50) NULL,
            voter_id VARCHAR(255) NULL,
            vote VARCHAR(50) NULL,
            meta_url TEXT NULL,
            comment TEXT NULL,
            processed_at TIMESTAMP WITH TIME ZONE NULL,
            created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
            updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
          )
        `);
        await client.query(`
          CREATE UNIQUE INDEX IF NOT EXISTS "idx_${md5Short}_voter_block_time" ON "${tableName}"(voter_id, block_time);
          CREATE INDEX IF NOT EXISTS "idx_${md5Short}_voter_id" ON "${tableName}"(voter_id);
          CREATE INDEX IF NOT EXISTS "idx_${md5Short}_voter_role" ON "${tableName}"(voter_role);
          CREATE INDEX IF NOT EXISTS "idx_${md5Short}_block_time" ON "${tableName}"(block_time DESC);
          CREATE INDEX IF NOT EXISTS "idx_${md5Short}_vote" ON "${tableName}"(vote)
        `);
        await client.query(
          'UPDATE proposals SET activities_table_name = $1, activities_table_created = TRUE WHERE proposal_id = $2',
          [tableName, pid]
        );
        created++;
      } catch (e) {
        error(`[proposals] Failed to create ga_* table for ${pid}: ${e.message}`);
      }
      process.stdout.write(`\r  [proposals] tables ${idx + 1}/${missing.length} created=${created}`);
    }
    info(`\n[proposals] Created ${created}/${missing.length} ga_* tables`);

    // 10. Verify
    const cntRes = await client.query('SELECT count(*)::int as cnt FROM proposals');
    info(`[proposals] Row count: ${cntRes.rows[0].cnt}`);
    return cntRes.rows[0].cnt;
  } finally {
    await client.end();
  }
}

if (require.main === module) {
  syncProposals().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { syncProposals };